definitions/npm/firebase_v4.x.x/flow_v0.46.x-/firebase_v4.x.x.js

```diff
@@ -0,0 +1,410 @@
+  delete(): Promise<void>;
+  getIdToken(forceRefresh?: boolean): Promise<string>;
+  getToken(forceRefresh?: boolean): Promise<string>;
+  linkAndRetrieveDataWithCredential(credential: AuthCredential): Promise<UserCredential>;
+  linkWithCredential(credential: AuthCredential): Promise<User>;
+  linkWithPhoneNumber(phoneNumber: string, applicationVerifier: ApplicationVerifier): Promise<ConfirmationResult>;
+  linkWithPopup(provider: OAuthProvider): Promise<UserCredential>;
+  reauthenticateAndRetrieveDataWithCredential(credential: AuthCredential): Promise<void>;
+  reauthenticateWithCredential(credential: AuthCredential): Promise<void>;
+  reauthenticateWithPhoneNumber(phoneNumber: string, applicationVerifier: ApplicationVerifier): Promise<ConfirmationResult>;
+  reload(): Promise<void>;
+  sendEmailVerification(): Promise<void>;
+  toJSON(): Object;
+  unlink(providerId: string): Promise<User>;
+  updateEmail(newEmail: string): Promise<void>;
+  updatePassword(newPassword: string): Promise<void>;
+  updatePhoneNumber(phoneCredential: AuthCredential): Promise<void>;
+  updateProfile(profile: { displayName: ?string, photoURL: ?string }): Promise<void>;
+}
+
+declare class EmailAuthProvider extends AuthProvider {
+  PROVIDER_ID: string;
+  providerId: string;
+  credential(email: string, password: string): AuthCredential;
+}
+
+declare class FacebookAuthProvider extends AuthProvider  {
+  PROVIDER_ID: string;
+  credential(token: string): AuthCredential;
+  addScope(scope: string): FacebookAuthProvider;
+  setCustomParameters(customOAuthParameters: Object): FacebookAuthProvider;
+}
+
+declare class GithubAuthProvider extends AuthProvider {
+  PROVIDER_ID: string;
+  credential(token: string): AuthCredential;
+  addScope(scope: string): GithubAuthProvider;
+  setCustomParameters(customOAuthParameters: Object): GithubAuthProvider;
+}
+
+declare class GoogleAuthProvider extends AuthProvider {
+  PROVIDER_ID: string;
+  credential(idToken?: string, accessToken?: string): AuthCredential;
+  addScope(scope: string): GoogleAuthProvider;
+  setCustomParameters(customOAuthParameters: Object): GoogleAuthProvider;
+}
+
+declare class PhoneAuthProvider extends AuthProvider {
+  PROVIDER_ID: string;
+  constructor(auth?: Auth): PhoneAuthProvider;
+  credential(verificationId: string, verificationCode: string): AuthCredential;
+  verifyPhoneNumber(phoneNumber: string, applicationVerifier: ApplicationVerifier): Promise<string>;
+}
+
+declare class TwitterAuthProvider extends AuthProvider {
+  PROVIDER_ID: string;
+  credential(token: string, secret: string): AuthCredential;
+  setCustomParameters(customOAuthParameters: Object): this;
+}
+
+declare type OAuthProvider = | FacebookAuthProvider
+                             | GithubAuthProvider
+                             | GoogleAuthProvider
+                             | TwitterAuthProvider;
+
+/****** database ******/
+declare type FirebaseValue = {} | Array<FirebaseValue> | string | number | boolean | null;
```

Comment:
I was looking how typescript people also set it as. It seems they made it any https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/firebase/index.d.ts#L20

Also working with nested values coming from val() makes it really hard to work with. Especially when you pass the values to different functions and want to be a type you specified in your project.